nomt/src/new_seek.rs

```diff
@@ -0,0 +1,421 @@
+        // children.
+        if !trie::is_leaf(&cur_node) {
+            self.page_id = Some(page_id);
+        }
+
+        // take enough bits to get us to the end of this page.
+        let bits = self.key.view_bits::<Msb0>()[self.position.depth() as usize..]
+            .iter()
+            .by_vals()
+            .take(DEPTH);
+
+        for bit in bits {
+            if trie::is_leaf(&cur_node) {
+                let children = if self.position.depth() as usize % DEPTH > 0 {
+                    self.position.child_node_indices()
+                } else {
+                    // note: the only time this is true in this loop is in the first iteration,
+                    // because we only take `DEPTH` bits.
+                    ChildNodeIndices::next_page()
+                };
+
+                self.state = RequestState::Completed(Some(trie::LeafData {
+                    key_path: page.node(&read_pass, children.left()),
+                    value_hash: page.node(&read_pass, children.right()),
+                }));
+                return;
+            }
+
+            self.position.down(bit);
+
+            cur_node = page.node(&read_pass, self.position.node_index());
+            if record_siblings {
+                self.siblings
+                    .push(page.node(&read_pass, self.position.sibling_index()));
+            }
+
+            if trie::is_terminator(&cur_node) {
+                self.state = RequestState::Completed(None);
+                return;
+            }
+
+            // leaf is handled in next iteration _or_ if this is the last iteration, in the next
+            // `continue_seek`.
+        }
+
+        self.state = RequestState::Seeking(cur_node);
+    }
+}
+
+enum RequestState {
+    Seeking(Node),
+    Completed(Option<trie::LeafData>),
+}
+
+enum SinglePageRequestState {
+    Pending(PageId),
+    Submitted,
+    Completed,
+}
+
+/// The `Seeker` seeks for the terminal nodes of multiple keys simultaneously, multiplexing requests
+/// over an I/O pool. Advance the seeker, provide new keys, and handle completions. A key path
+/// request is considered completed when the terminal node has been found, and all pages along the
+/// path to the terminal node are in the page cache. This includes the page that contains the leaf
+/// children of the request.
+///
+/// While requests are normally completed in the order they're submitted,
+/// it is possible to submit a special request for a single page, which will be prioritized over
+/// other requests and whose completion will be handled first. This has a particular use case in
+/// mind: the fetching of sibling leaf node child pages, as is an edge case during merkle tree
+/// updates.
+pub struct Seeker {
+    root: Node,
+    page_cache: PageCache,
+    page_loader: PageLoader,
+    processed: usize,
+    requests: VecDeque<(bool, SeekRequest)>,
+    page_loads: HashMap<PageId, Vec<usize>>,
+    page_load_slab: Slab<PageLoad>,
+    slab_retries: VecDeque<usize>,
```

Comment:
would be good to leave a comment on the ordering, i.e. when exactly we push in front vs. back.